osquery/tables/applications/browser_utils.cpp

```diff
@@ -174,58 +260,116 @@ void genExtension(const std::string& uid,
+  } catch (const pt::json_parser::json_parser_error& /* e */) {
+    VLOG(1) << "Could not parse JSON from: " << path + kManifestFile;
+    return;
+  }
+
+  const std::string& version = tree.get<std::string>("version", "");
+  auto& scriptMatchPairs = contentScriptMap[std::make_tuple(
+      fs::path(path).parent_path().parent_path().leaf().string(), version)];
+
+  auto contentScriptDetail = genContentScriptDetail(tree);
+  for (auto& contentScript : contentScriptDetail) {
+    for (auto& script : contentScript[kScriptKey]) {
+      if (contentScript[kMatchesKey].empty()) {
+        scriptMatchPairs.insert(std::make_tuple(script, ""));
+      } else {
+        for (auto& match : contentScript[kMatchesKey]) {
+          scriptMatchPairs.insert(std::make_tuple(script, match));
+        }
+      }
+    }
+  }
+}
+
 QueryData genChromeBasedExtensions(QueryContext& context,
                                    const std::vector<fs::path>& chromePaths) {
   QueryData results;
 
-  auto users = usersFromContext(context);
-  for (const auto& row : users) {
-    if (row.count("uid") > 0 && row.count("directory") > 0) {
-      // For each user, enumerate all of their chrome profiles.
-      std::vector<std::string> profiles;
-      for (const auto& chromePath : chromePaths) {
-        fs::path extension_path = row.at("directory") / chromePath;
-        if (!resolveFilePattern(extension_path, profiles, GLOB_FOLDERS).ok()) {
-          continue;
+  const auto& extensionPathsByUser =
+      chromeExtensionPathsByUser(usersFromContext(context), chromePaths);
+
+  for (const auto& userExtensionPaths : extensionPathsByUser) {
+    const auto& uid = std::get<0>(userExtensionPaths);
+    std::map<fs::path, std::string> profileNameMap;
+
+    for (const auto& version : std::get<1>(userExtensionPaths)) {
+      const auto& profile_path = fs::path(version)
+                                     .parent_path()
+                                     .parent_path()
+                                     .parent_path()
+                                     .parent_path();
+
+      auto it = profileNameMap.find(profile_path);
+      if (it == profileNameMap.end()) {
+        auto status =
+            getChromeProfileName(profileNameMap[profile_path], profile_path);
+        if (!status.ok()) {
+          LOG(WARNING) << "Getting Chrome profile name failed: "
+                       << status.getMessage();
         }
+      }
 
-        // For each profile list each extension in the Extensions directory.
-        for (const auto& profile : profiles) {
-          std::vector<std::string> extensions = {};
-          listDirectoriesInDirectory(profile, extensions);
+      genExtension(uid, version, profileNameMap[profile_path], results);
+    }
+  }
 
-          if (extensions.empty()) {
-            continue;
-          }
+  return results;
+}
 
-          auto profile_path = fs::path(profile).parent_path().parent_path();
+QueryData genChromeBasedExtensionContentScripts(
```

Comment:
This implementation will always scan all extensions for the requested users. This might be good enough. But you might be able to make this faster if you expect JOINs to happen using an `identifier` and if you can construct the expected path to the extension from the user + identifier?

If that is possible then add `index=True` to `identifier` and allow the `chromeExtensionPathsByUser` method to take the input identifier from the context.

Reply:
If you are confident that you must scan all of a user's extensions (because you cannot predict the path from the identifier) then the current implementation is OK.

Reply:
Yeah, I unfortunately don't think we can predict the path from the identifier -- a single extension ID can identify multiple installs within a single Chrome profile (different versions), multiple installs across a single Chrome install (different profiles), or even multiple installs across multiple Chrome installs (Chrome + Chromium + Chromium derivatives) 😞